### ISSUE TITLE
Endpoint in hostNetwork of LoadBalancer, NodePort, externalIP Service cannot be reached from outside cluster

### ISSUE THREAD

**Issue by maxbischoff** [cilium/cilium#10402]
## Bug report

**General Information**

- Cilium version (run `cilium version`)
`Client: 1.7.0 adeaf8c04 2020-02-18T21:41:10+01:00 go version go1.13.8 linux/amd64
Daemon: 1.7.0 adeaf8c04 2020-02-18T21:41:10+01:00 go version go1.13.8 linux/amd64`
- Kernel version (run `uname -a`)
`4.19.86-flatcar`
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...) 
`kubectl version`
`Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7", GitCommit:"be3d344ed06bff7a4fc60656200a93c74f31f9a4", GitTreeState:"clean", BuildDate:"2020-02-11T19:24:46Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}`
- Link to relevant artifacts (policies, deployments scripts, ...)
We deployed the helm chart from https://github.com/maxbischoff/cilium/tree/patches/install/kubernetes/cilium with following values:
<details><summary>values.yaml</summary>
<p>

```yaml
global:
  tag: v1.7.0
  psp:
    enabled: true
  kubeProxyReplacement: "strict"
  hostServices:
    enabled: true
  k8sServiceHost: "127.0.0.1"
  k8sServicePort: 11000
  nodePort:
    enabled: true
  externalIPs:
    enabled: true
agent:
  resources:
    requests:
      cpu: 500m
      memory: 500Mi
      ephemeral-storage: 250Mi
  monitor:
    resources:
      requests:
        cpu: 500m
        memory: 500Mi
        ephemeral-storage: 250Mi
operator:
  resources:
    requests:
      cpu: 500m
      memory: 500Mi
      ephemeral-storage: 250Mi
```

</p>
</details>

kube-proxy is **not** deployed
- Upload a system dump (run `curl -sLO
https://github.com/cilium/cilium-sysdump/releases/latest/download/cilium-sysdump.zip &&
python cilium-sysdump.zip` and then attach the generated zip file)
[cilium-sysdump-20200302-110411.zip](https://github.com/cilium/cilium/files/4275451/cilium-sysdump-20200302-110411.zip)

**How to reproduce the issue**

1. Create a pod running in the host Network
2. expose it with a loadBalancer service with `externalTrafficPolicy: Local`
3. curl the service, observe that the traffic times out

**Comment by brb** [comment]
- Did you curl the LoadBalancer svc from outside or from within the cluster?
- Can you attach the sysdump, or at least get the output of `cilium bpf lb list`?

**Comment by maxbischoff** [comment]
* I curled the LoadBalancer svc from outside the cluster
* I can at least provide that output, I'll have to check if I can also provide the sysdump. 
```
kubectl exec -it cilium-fxw27 -- cilium bpf lb list
SERVICE ADDRESS       BACKEND ADDRESS                       
100.72.178.101:80     100.64.5.159:8000 (14)                
                      100.64.3.128:8000 (14)                
                      0.0.0.0:0 (14) [ClusterIP]            
100.64.5.14:30157     100.64.5.143:8443 (30)                
                      100.64.3.15:8443 (30)                 
                      0.0.0.0:0 (30) [NodePort]             
10.88.58.5:443        10.234.32.131:11000 (3)               
                      10.234.32.78:11000 (3)                
                      10.234.32.132:11000 (3)               
                      10.234.32.77:11000 (3)                
                      10.234.32.76:11000 (3)                
                      10.234.32.75:11000 (3)                
                      0.0.0.0:0 (3) [ExternalIPs]           
100.72.65.50:443      100.64.5.172:8443 (22)                
                      0.0.0.0:0 (22) [ClusterIP]            
100.72.76.202:8080    100.64.5.205:8080 (19)                
                      0.0.0.0:0 (19) [ClusterIP]            
100.72.0.1:443        10.234.32.75:10001 (8)                
                      0.0.0.0:0 (8) [ClusterIP]             
                      10.234.32.77:10001 (8)                
                      10.234.32.76:10001 (8)                
0.0.0.0:32432         100.64.3.15:8000 (31)                 
                      0.0.0.0:0 (31) [NodePort]             
                      100.64.5.143:8000 (31)                
10.234.32.132:32166   0.0.0.0:0 (12) [NodePort]             
                      100.64.5.159:8000 (12)                
                      100.64.3.128:8000 (12)                
100.72.109.195:80     0.0.0.0:0 (10) [ClusterIP]            
                      100.64.3.128:8000 (10)                
                      100.64.5.159:8000 (10)                
100.64.5.14:31385     100.64.3.128:8000 (16)                
                      0.0.0.0:0 (16) [NodePort]             
                      100.64.5.159:8000 (16)                
100.72.235.239:80     0.0.0.0:0 (25) [ClusterIP]            
                      100.64.3.15:8080 (25)                 
                      100.64.5.143:8080 (25)                
100.72.33.21:80       0.0.0.0:0 (23) [ClusterIP]            
                      100.64.4.149:9090 (23)                
                      100.64.5.157:9090 (23)                
100.72.134.197:80     0.0.0.0:0 (43) [ClusterIP]            
                      100.64.5.1:8080 (43)                  
                      100.64.3.10:8080 (43)                 
10.234.32.132:30880   0.0.0.0:0 (38) [NodePort, Local]      
                      100.64.5.1:8000 (38)                  
100.64.5.14:30880     0.0.0.0:0 (39) [NodePort, Local]      
                      100.64.5.1:8000 (39)                  
100.64.5.14:32432     100.64.3.15:8000 (33)                 
                      0.0.0.0:0 (33) [NodePort]             
                      100.64.5.143:8000 (33)                
100.72.95.117:8000    100.64.4.54:8000 (21)                 
                      0.0.0.0:0 (21) [ClusterIP]            
10.88.58.14:443       100.64.5.1:8443 (44)                  
                      0.0.0.0:0 (44) [ExternalIPs, Local]   
100.64.5.14:32166     100.64.5.159:8000 (13)                
                      0.0.0.0:0 (13) [NodePort]             
                      100.64.3.128:8000 (13)                
0.0.0.0:31385         0.0.0.0:0 (17) [NodePort]             
                      100.64.3.128:8000 (17)                
                      100.64.5.159:8000 (17)                
10.234.32.132:30157   100.64.5.143:8443 (29)                
                      0.0.0.0:0 (29) [NodePort]             
                      100.64.3.15:8443 (29)                 
0.0.0.0:30567         10.234.32.78:11000 (4)                
                      10.234.32.132:11000 (4)               
                      10.234.32.76:11000 (4)                
                      10.234.32.131:11000 (4)               
                      10.234.32.75:11000 (4)                
                      0.0.0.0:0 (4) [NodePort]              
                      10.234.32.77:11000 (4)                
0.0.0.0:32166         100.64.3.128:8000 (11)                
                      0.0.0.0:0 (11) [NodePort]             
                      100.64.5.159:8000 (11)                
100.72.123.85:443     100.64.3.15:8443 (26)                 
                      0.0.0.0:0 (26) [ClusterIP]            
                      100.64.5.143:8443 (26)                
10.234.32.132:30712   100.64.5.1:8443 (41)                  
                      0.0.0.0:0 (41) [NodePort, Local]      
0.0.0.0:30712         100.64.5.1:8443 (40)                  
                      0.0.0.0:0 (40) [NodePort, Local]      
100.72.220.209:80     100.64.5.242:8080 (18)                
                      0.0.0.0:0 (18) [ClusterIP]            
100.72.0.10:53        100.64.3.51:53 (1)                    
                      0.0.0.0:0 (1) [ClusterIP]             
                      100.64.3.150:53 (1)                   
                      100.64.3.135:53 (1)                   
100.72.123.85:80      0.0.0.0:0 (27) [ClusterIP]            
                      100.64.5.143:8000 (27)                
                      100.64.3.15:8000 (27)                 
100.72.2.112:443      100.64.3.10:8443 (36)                 
                      100.64.5.1:8443 (36)                  
                      0.0.0.0:0 (36) [ClusterIP]            
10.234.32.132:30567   10.234.32.131:11000 (5)               
                      10.234.32.78:11000 (5)                
                      10.234.32.76:11000 (5)                
                      10.234.32.132:11000 (5)               
                      10.234.32.77:11000 (5)                
                      0.0.0.0:0 (5) [NodePort]              
                      10.234.32.75:11000 (5)                
100.72.92.79:9115     100.64.4.26:9115 (9)                  
                      0.0.0.0:0 (9) [ClusterIP]             
10.234.32.132:32432   0.0.0.0:0 (32) [NodePort]             
                      100.64.5.143:8000 (32)                
                      100.64.3.15:8000 (32)                 
100.72.7.131:443      0.0.0.0:0 (7) [ClusterIP]             
                      100.64.5.23:9100 (7)                  
100.72.227.73:443     100.64.5.56:8443 (20)                 
                      0.0.0.0:0 (20) [ClusterIP]            
100.72.9.109:80       100.64.5.21:9090 (34)                 
                      0.0.0.0:0 (34) [ClusterIP]            
100.64.5.14:30712     100.64.5.1:8443 (42)                  
                      0.0.0.0:0 (42) [NodePort, Local]      
10.88.58.14:80        0.0.0.0:0 (45) [ExternalIPs, Local]   
                      100.64.5.1:8000 (45)                  
100.72.104.35:443     0.0.0.0:0 (2) [ClusterIP]             
                      10.234.32.132:11000 (2)               
                      10.234.32.131:11000 (2)               
                      10.234.32.77:11000 (2)                
                      10.234.32.76:11000 (2)                
                      10.234.32.78:11000 (2)                
                      10.234.32.75:11000 (2)                
0.0.0.0:30157         0.0.0.0:0 (28) [NodePort]             
                      100.64.3.15:8443 (28)                 
                      100.64.5.143:8443 (28)                
100.64.5.14:30567     10.234.32.75:11000 (6)                
                      10.234.32.77:11000 (6)                
                      10.234.32.132:11000 (6)               
                      10.234.32.78:11000 (6)                
                      10.234.32.131:11000 (6)               
                      10.234.32.76:11000 (6)                
                      0.0.0.0:0 (6) [NodePort]              
100.72.168.8:80       100.64.5.190:9090 (24)                
                      100.64.3.75:9090 (24)                 
                      0.0.0.0:0 (24) [ClusterIP]            
0.0.0.0:30880         0.0.0.0:0 (37) [NodePort, Local]      
                      100.64.5.1:8000 (37)                  
100.72.2.112:80       100.64.5.1:8000 (35)                  
                      100.64.3.10:8000 (35)                 
                      0.0.0.0:0 (35) [ClusterIP]            
10.234.32.132:31385   100.64.3.128:8000 (15)                
                      100.64.5.159:8000 (15)                
                      0.0.0.0:0 (15) [NodePort]   
```

The service in question is:
```
kubectl get svc kubernetes-api -o wide
NAME             TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)         AGE     SELECTOR
kubernetes-api   LoadBalancer   100.72.104.35   10.88.58.5    443:30567/TCP   2d19h   app=system-haproxy
```

**Comment by brb** [comment]
Thanks. Can you check whether the svc is accessible via NodePort (30567)?

**Comment by maxbischoff** [comment]
I cannot reach the service on its NodePort.

**Comment by brb** [comment]
Can you reach the svc from inside the cluster, e.g. by using `127.0.0.1:30567`?

**Comment by maxbischoff** [comment]
I can reach the service on both nodes of the cluster as well as pods in the cluster using `curl 127.0.0.1:30567`

**Comment by brb** [comment]
Can you privately send me the sysdump (I'm `martynas` on the Cilium's slack)?

**Comment by maxbischoff** [comment]
I just attached my sysdump https://github.com/cilium/cilium/files/4275451/cilium-sysdump-20200302-110411.zip

**Comment by brb** [comment]
This is the case where we don't apply the rev-DNAT xlation for a reply from a service as the reply from a pod in host netns is not handled by bpf_lxc.o. I'm currently working on the fix.

**Comment by kstiehl** [comment]
I run into the same problem

**Comment by mmack** [comment]
same here

**Comment by zhangjianghua** [comment]
1.8.2 also encountered this problem；

**Comment by zhangjianghua** [comment]
1.8.2 also encountered this problem；
If the service sets externalTrafficPolicy: Local, it will not work
If the service is set to externalTrafficPolicy: Cluster, it can be connected, but when the pod ip and node ip are the same, the request will time out
Node access within the cluster is always good

**Comment by borkmann** [comment]
> 1.8.2 also encountered this problem；
> If the service sets externalTrafficPolicy: Local, it will not work
> If the service is set to externalTrafficPolicy: Cluster, it can be connected, but when the pod ip and node ip are the same, the request will time out
> Node access within the cluster is always good

Yeah, that is a known limitation to Pods in host Network. We're currently working to find an efficient solution to address this corner case w/o adding too much overhead to the regular fast-path. Workaround for now would be to move them out of host Network.

**Comment by brb** [comment]
This is going to be fixed by https://github.com/cilium/cilium/pull/17504.

**Comment by brb** [comment]
Assigning myself to document the new flag.

**Comment by brb** [comment]
> any eta for that?

It's blocked by https://github.com/cilium/cilium/pull/17504.

**Comment by mhulscher** [comment]
Currently blocked by this as well. Was considering running my NLB backend pods (ingress) w/ `hostNetwork: true`, considering that the pods need an IP-address that lives on the primary NIC, as a workaround for this issue: https://github.com/cilium/cilium/issues/19250

**Comment by brb** [comment]
https://github.com/cilium/cilium/pull/22978 should fix this issue.

**Comment by julianwiedmann** [comment]
> #22978 should fix this issue.

yup. Note that `hostNetwork:true` with `eTP:local` (that's the initial config in this issue) should already be fixed with https://github.com/cilium/cilium/pull/22756.

**Comment by julianwiedmann** [comment]
> > #22978 should fix this issue.
> 
> yup. Note that `hostNetwork:true` with `eTP:local` (that's the initially reported config in this issue) should already be fixed with #22756.

Actually let's close this issue based on that :).

The DSR aspect lives in https://github.com/cilium/cilium/issues/10789.
